quinn-proto/src/connection/mod.rs

```diff
@@ -2519,17 +2693,29 @@ where
 
     /// Returns Err(()) if no CIDs were available
     fn update_rem_cid(&mut self) -> Result<(), ()> {
-        let (cid, retired) = self.rem_cids.next().ok_or(())?;
-        trace!("switching to remote CID {}: {}", cid.sequence, cid.id);
-
-        // Retire the current remote CID and any CIDs we had to skip.
+        // Retire the current remote CID
         let retire_cids = &mut self.spaces[SpaceId::Data].pending.retire_cids;
         retire_cids.push(self.rem_cid_seq);
-        retire_cids.extend(retired);
 
-        // Apply the new CID
+        match self.rem_cids.next() {
+            Some(cid) => {
+                trace!("switching to remote CID {}: {}", cid.0.sequence, cid.0.id);
+
+                // Retire any CIDs we had to skip.
+                retire_cids.extend(cid.1);
+
+                // Apply new remote CID
+                self.apply_new_rem_cid(cid.0);
+                Ok(())
+            }
+            None => Err(()),
+        }
+    }
```

Comment:
By moving the retiring of the current CID before checking whether we have any others, this change removes the check that prevents us from retiring a remote CID when don't have any others to switch to, potentially rendering the connection unusable. Is that intended?

Reply:
I see your point.  This situation was (partially) considered when processing NEW_CONNECTION_ID frame. If we don't have any other CIDs to switch to, `fn update_rem_cid ` return `Err(())` and will be caught by `match valid_rem_cid` where the current CID in the frame will be activated.  However, i do observe now that  `fn update_rem_cid` is used in multiple places (e.g., after migration) which could lead the situation as you mentioned. Working on a fix now.

Reply:
Also, may I ask the design motivation of having a separate `rem_cid` in Connection struct, instead of having all remote CIDs managed by `CidQueue`? This makes CID management kind of complicated when dealing with retirement.

Reply:
No strong reason. If bundling them together makes things simpler/cleaner I'd be happy to see that happen.